<a href="https://colab.research.google.com/github/swetha097/FAMBench/blob/main/Vicuna_Working_copy_of_nvidia_tensorrt.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://colab.research.google.com/github/run-llama/llama_index/blob/main/docs/docs/examples/llm/nvidia_tensorrt.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!which python
!nvidia-smi
!which pip
!python --version
!apt-get update && apt-get -y install python3.10

/usr/local/bin/python
Thu Jan 30 04:40:38 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   42C    P8              9W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-------------------------

In [ ]:
!python --version
!python3 --version
!which python
!which python3
!which python3.10

Python 3.11.11
Python 3.11.11
/usr/local/bin/python
/usr/bin/python3
/usr/bin/python3.10


In [ ]:
!ln -sf /usr/bin/python3.10 /usr/local/bin/python
!ln -sf /usr/bin/python3.10 /usr/local/bin/python3
!ln -sf /usr/bin/python3.10 /usr/bin/python3
!which python
!apt install python3-pip openmpi-bin libopenmpi-dev git git-lfs wget

/usr/local/bin/python
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
libopenmpi-dev is already the newest version (4.1.2-2ubuntu1).
openmpi-bin is already the newest version (4.1.2-2ubuntu1).
openmpi-bin set to manually installed.
git is already the newest version (1:2.34.1-1ubuntu1.12).
wget is already the newest version (1.21.2-2ubuntu1.1).
git-lfs is already the newest version (3.0.2-1ubuntu0.3).
The following additional packages will be installed:
  python3-setuptools python3-wheel
Suggested packages:
  python-setuptools-doc
The following NEW packages will be installed:
  python3-pip python3-setuptools python3-wheel
0 upgraded, 3 newly installed, 0 to remove and 22 not upgraded.
Need to get 1,677 kB of archives.
After this operation, 8,968 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 python3-setuptools all 59.6.0-1.2ubuntu0.22.04.2 [340 kB]
Get:2 http://archive.ubuntu.com/

In [ ]:
!which pip3
!pip --version

/usr/local/bin/pip3
pip 22.0.2 from /usr/lib/python3/dist-packages/pip (python 3.10)


In [ ]:
!python -m pip --version

pip 22.0.2 from /usr/lib/python3/dist-packages/pip (python 3.10)


In [ ]:
!which python3
!which python

/usr/local/bin/python3
/usr/local/bin/python


In [ ]:
!python3 -m pip install tensorrt_llm==0.16 --extra-index-url https://pypi.nvidia.com

Looking in indexes: https://pypi.org/simple, https://pypi.nvidia.com
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 GB 29.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.5/73.5 KB 3.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 287.5/287.5 KB 20.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 536.2/536.2 KB 53.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.2/98.2 KB 12.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 111.0/111.0 KB 16.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.8/57.8 KB 210.2 kB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 868.8/868.8 KB 32.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.3/62.3 KB 10.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 389.6/389.6 KB 22.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.0/84.0 KB 12.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [ ]:
!python3 -m pip install llama-index-llms-nvidia-tensorrt

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 26.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 82.8/82.8 KB 11.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 59.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 55.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 63.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 301.8/301.8 KB 41.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 599.5/599.5 KB 47.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.8/50.8 KB 6.9 MB/s eta 0:00:00


In [ ]:
!pip install pynvml==11.5.0 --force-reinstall

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.1/53.1 KB 1.9 MB/s eta 0:00:00
  Attempting uninstall: pynvml
    Found existing installation: pynvml 12.0.0
    Uninstalling pynvml-12.0.0:
      Successfully uninstalled pynvml-12.0.0


In [ ]:
# from google.colab import drive
# drive.mount('/content/drive')

# !ls /content/drive
!git clone --branch v0.16.0 https://github.com/NVIDIA/TensorRT-LLM.git
!cd TensorRT-LLM/examples/gpt/ && pip install -r requirements.txt

Cloning into 'TensorRT-LLM'...
remote: Enumerating objects: 34420, done.
remote: Counting objects: 100% (17/17), done.
remote: Compressing objects: 100% (14/14), done.
remote: Total 34420 (delta 4), reused 3 (delta 3), pack-reused 34403 (from 2)
Receiving objects: 100% (34420/34420), 461.56 MiB | 16.00 MiB/s, done.
Resolving deltas: 100% (24829/24829), done.
Note: switching to '42a7b0922fc9e095f173eab9a7efa0bcdceadd0d'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

Updating files: 100% (3149/3149), done.
Filtering content: 

In [ ]:
!ls



sample_data  TensorRT-LLM


In [ ]:
!git lfs install
!git clone --branch master https://huggingface.co/swetha097/vicuna_model

Git LFS initialized.
Cloning into 'vicuna_model'...
remote: Enumerating objects: 16, done.
remote: Counting objects: 100% (13/13), done.
remote: Compressing objects: 100% (11/11), done.
remote: Total 16 (delta 0), reused 0 (delta 0), pack-reused 3 (from 1)
Unpacking objects: 100% (16/16), 222.71 KiB | 1000.00 KiB/s, done.
Filtering content: 100% (3/3), 5.11 GiB | 2.40 MiB/s, done.
Encountered 3 file(s) that may not have been copied correctly on Windows:
	pytorch_model-00003-of-00003.bin
	pytorch_model-00001-of-00003.bin
	pytorch_model-00002-of-00003.bin

See: `git lfs help smudge` for more details.


In [ ]:
!ls
!python3 -c "import tensorrt_llm"
!du -h
# !cd vicuna_model/ && git count-objects -vH && git lfs fetch --all && git lfs ls-files

sample_data  TensorRT-LLM  vicuna_model
[TensorRT-LLM] TensorRT-LLM version: 0.16.0
80K	./.config/logs/2025.01.28
84K	./.config/logs
8.0K	./.config/configurations
140K	./.config
44K	./TensorRT-LLM/scripts
12K	./TensorRT-LLM/.github/workflows
12K	./TensorRT-LLM/.github/ISSUE_TEMPLATE
28K	./TensorRT-LLM/.github
36K	./TensorRT-LLM/examples/bindings/executor
40K	./TensorRT-LLM/examples/bindings
60K	./TensorRT-LLM/examples/llm-api
16K	./TensorRT-LLM/examples/draft_target_model
32K	./TensorRT-LLM/examples/quantization
20K	./TensorRT-LLM/examples/cpp_library
32K	./TensorRT-LLM/examples/deepseek_v2
16K	./TensorRT-LLM/examples/python_plugin/plugin_lib
36K	./TensorRT-LLM/examples/python_plugin
40K	./TensorRT-LLM/examples/eagle
8.0K	./TensorRT-LLM/examples/blip2
20K	./TensorRT-LLM/examples/model_api
88K	./TensorRT-LLM/examples/sdxl
52K	./TensorRT-LLM/examples/recurrentgemma
28K	./TensorRT-LLM/examples/commandr
16K	./TensorRT-LLM/examples/nemotron
52K	./TensorRT-LLM/examples/qwen
128K	./TensorRT-L

1. Convert checkpoints to a format accepted by tensorrt-llm

In [ ]:
!mkdir TensorRT-LLM/tllm_checkpoint_1gpu_fp16
!rm -r vicuna_model/.git/
!ls
!du -h


mkdir: cannot create directory ‘TensorRT-LLM/tllm_checkpoint_1gpu_fp16’: File exists
sample_data  TensorRT-LLM  vicuna_model
80K	./.config/logs/2025.01.28
84K	./.config/logs
8.0K	./.config/configurations
140K	./.config
4.0K	./TensorRT-LLM/tllm_checkpoint_1gpu_fp16
44K	./TensorRT-LLM/scripts
12K	./TensorRT-LLM/.github/workflows
12K	./TensorRT-LLM/.github/ISSUE_TEMPLATE
28K	./TensorRT-LLM/.github
36K	./TensorRT-LLM/examples/bindings/executor
40K	./TensorRT-LLM/examples/bindings
60K	./TensorRT-LLM/examples/llm-api
16K	./TensorRT-LLM/examples/draft_target_model
32K	./TensorRT-LLM/examples/quantization
20K	./TensorRT-LLM/examples/cpp_library
32K	./TensorRT-LLM/examples/deepseek_v2
16K	./TensorRT-LLM/examples/python_plugin/plugin_lib
36K	./TensorRT-LLM/examples/python_plugin
40K	./TensorRT-LLM/examples/eagle
8.0K	./TensorRT-LLM/examples/blip2
20K	./TensorRT-LLM/examples/model_api
88K	./TensorRT-LLM/examples/sdxl
52K	./TensorRT-LLM/examples/recurrentgemma
28K	./TensorRT-LLM/examples/commandr


In [ ]:
 !python3 TensorRT-LLM/examples/llama/convert_checkpoint.py --model_dir ./vicuna_model/ --output_dir ./TensorRT-LLM/tllm_checkpoint_1gpu_fp16/

[TensorRT-LLM] TensorRT-LLM version: 0.16.0
0.16.0
[01/30/2025-06:24:28] [TRT-LLM] [I] Specified dtype 'auto'; inferred dtype 'float16'.
[01/30/2025-06:24:28] [TRT-LLM] [W] Implicitly setting LLaMAConfig.tie_word_embeddings = False
192it [01:52,  1.23it/s]^C


In [ ]:
!ls TensorRT-LLM/tllm_checkpoint_1gpu_fp16

In [ ]:
!python3 -c "from llama_index.llms.nvidia_tensorrt import LocalTensorRTLLM; llm = LocalTensorRTLLM(model_path='./engine_outputs', engine_name='gpt_float16_tp1_rank0.engine', tokenizer_dir='gpt2', max_new_tokens=40)"

[TensorRT-LLM] TensorRT-LLM version: 0.16.0
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.10/dist-packages/llama_index/llms/nvidia_tensorrt/base.py", line 170, in __init__
    raise ValueError(
ValueError: Provided model path does not exist. Please check the path or provide a model_url to download.


# Nvidia TensorRT-LLM


TensorRT-LLM provides users with an easy-to-use Python API to define Large Language Models (LLMs) and build TensorRT engines that contain state-of-the-art optimizations to perform inference efficiently on NVIDIA GPUs.

[TensorRT-LLM Github](https://github.com/NVIDIA/TensorRT-LLM)

## TensorRT-LLM Environment Setup
Since TensorRT-LLM is a SDK for interacting with local models in process there are a few environment steps that must be followed to ensure that the TensorRT-LLM setup can be used. Please note, that Nvidia Cuda 12.2 or higher is currently required to run TensorRT-LLM.

In this tutorial we will show how to use the connector with GPT2 model.
For the best experience, we recommend following
[Installation](https://github.com/NVIDIA/TensorRT-LLM/tree/v0.8.0?tab=readme-ov-file#installation) process on the
official [TensorRT-LLM Github](https://github.com/NVIDIA/TensorRT-LLM).

The following steps are showing how to set up your model with TensorRT-LLM v0.8.0 for x86_64 users.

1. Obtain and start the basic docker image environment.
```
docker run --rm --runtime=nvidia --gpus all --entrypoint /bin/bash -it nvidia/cuda:12.1.0-devel-ubuntu22.04
```

2. Install dependencies, TensorRT-LLM requires Python 3.10
```
apt-get update && apt-get -y install python3.10 python3-pip openmpi-bin libopenmpi-dev git git-lfs wget
```
3. Install the latest stable version (corresponding to the release branch) of TensorRT-LLM. We are using version 0.8.0, but for the most up to date release,
please refer to [official release page](https://github.com/NVIDIA/TensorRT-LLM/releases).
```
pip3 install tensorrt_llm==0.8.0 -U --extra-index-url https://pypi.nvidia.com
```

4. Check installation
```
python3 -c "import tensorrt_llm"
```
The above command should not produce any errors.

5. For this example we will use GPT2. The GPT2 model files need to be created via scripts following the instructions [here](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples/gpt#usage)
    * First, inside the container, we've started during stage 1, clone TensorRT-LLM repository:
    ```
    git clone --branch v0.8.0 https://github.com/NVIDIA/TensorRT-LLM.git
    ```
    * Install requirements for GPT2 model with:
    ```
    cd TensorRT-LLM/examples/gpt/ && pip install -r requirements.txt
    ```
    * Download hf gpt2 model
    ```
    rm -rf gpt2 && git clone https://huggingface.co/gpt2-medium gpt2
    cd gpt2
    rm pytorch_model.bin model.safetensors
    wget -q https://huggingface.co/gpt2-medium/resolve/main/pytorch_model.bin
    cd ..
    ```
    * Convert weights from HF Transformers to TensorRT-LLM format
    ```
    python3 hf_gpt_convert.py -i gpt2 -o ./c-model/gpt2 --tensor-parallelism 1 --storage-type float16
    ```
    * Build TensorRT engine
    ```
    python3 build.py --model_dir=./c-model/gpt2/1-gpu --use_gpt_attention_plugin --remove_input_padding
    ```
  
6. Install `llama-index-llms-nvidia-tensorrt` package
  ```
  pip install llama-index-llms-nvidia-tensorrt
  ```

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

!ls /content/drive

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
MyDrive


## Basic Usage

#### Call `complete` with a prompt

```python
from llama_index.llms.nvidia_tensorrt import LocalTensorRTLLM

llm = LocalTensorRTLLM(
    model_path="./engine_outputs",
    engine_name="gpt_float16_tp1_rank0.engine",
    tokenizer_dir="gpt2",
    max_new_tokens=40,
)

resp = llm.complete("Who is Harry Potter?")
print(str(resp))
```

The expected response should look like:
```
Harry Potter is a fictional character created by J.K. Rowling in her first novel, Harry Potter and the Philosopher's Stone. The character is a wizard who lives in the fictional town#
```